Firestore/core/src/firebase/firestore/util/filesystem_win.cc

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/util/filesystem.h"
+
+#include <windows.h>
+
+#include <cerrno>
+#include <string>
+
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "Firestore/core/src/firebase/firestore/util/path.h"
+#include "Firestore/core/src/firebase/firestore/util/string_format.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+bool Exists(const Path& path, bool* is_directory) {
+  DWORD attrs = ::GetFileAttributesW(path.c_str());
+  if (attrs == INVALID_FILE_ATTRIBUTES) {
+    return false;
+  }
+
+  if (is_directory) {
+    *is_directory = attrs & FILE_ATTRIBUTE_DIRECTORY;
+  }
+  return true;
+}
+
+Path TempDir() {
+  // Returns a null-terminated string with a trailing backslash.
+  wchar_t buffer[MAX_PATH + 1];
+  DWORD count = ::GetTempPathW(MAX_PATH, buffer);
+  HARD_ASSERT(count > 0, "Failed to determine temporary directory %s",
+              ::GetLastError());
+
```

Comment:
I recommend adding at least:
```
HARD_ASSERT(count <= MAX_PATH, "Temporary directory doesn't fit in MAX_PATH characters.");
```

IIUC, it's possible for this GetTempPathW call to return MAX_PATH+1 in cases where the temp path doesn't fit within MAX_PATH wchars.

Also note that (apparently) if the user has prepended \? to their tmp path, then this path could be up to 32k wide chars long. (And you can even drop the \? requirement on windows 10+ and opt-in to that being the default.)


You could alternatively allocate 32k (bleah) and use that. (You'd still want to do the hard_assert; but it should be impossible for that to trigger unless the user sets an exceptionally bad tmp path on purpose to see what breaks.)